File: src/learning_gem5/my_object.cc
#include "learning_gem5/my_object.hh"
#include "debug/MyDebug.hh"
#include <iostream>
#include <string>

MyObject::MyObject(MyObjectParams *params): SimObject(params),
myEvent([this]{processEvent();},name())
{ 
    this->latency     = params->time_to_wait;
    this->timesLeft   = params->number_of_fires;
    this->myByeObject = params->bye_bye_object;
    this->myName      = params->name;

    std::cout << "My World! From a SimObject" <<std::endl;
    DPRINTF(MyDebug, "Debug Flag Check\n");
}

MyObject* MyObjectParams::create()
{
    return new MyObject(this);
}

void MyObject::processEvent()
{
    timesLeft--;

    if(timesLeft == 0 )
    {
	DPRINTF(MyDebug, "Done with my events!!\n");
	(*myByeObject).sayByeBye(myName);
    }
    else
    {
	DPRINTF(MyDebug, "Hi! I'm processing an event, %d left\n", timesLeft );
	schedule( myEvent, curTick()+latency );
    };
}

void MyObject::startup()
{
    schedule(myEvent, 200);
}

File: src/learning_gem5/MyObject.py
from m5.params import *
from m5.SimObject import SimObject

class ByeByeObject(SimObject):
    type = 'ByeByeObject'
    cxx_header = "learning_gem5/byebye_object.hh"

    buffer_size     = Param.MemorySize('1kB', "Size of buffer to write good Byes")
    write_bandwidth = Param.MemoryBandwidth('100MB/s', "Bandwidth to fill the buffer")

class MyObject(SimObject):
    type = 'MyObject'
    cxx_header = "learning_gem5/my_object.hh"

    time_to_wait    = Param.Latency('2us', "Time before firing the event" )
    number_of_fires = Param.Int(5, "Number of times to fire the event before good bye" )
    bye_bye_object  = Param.ByeByeObject("A bye bye object")
    name	    = Param.String("Name of the owner")


